site/docs/expressions/scalar_functions.md

```diff
@@ -43,10 +43,68 @@ For scalar functions (one value produced for each record), a function binding de
+
+### Direct Return Types
+
+A direct return type is one that is fully known at function definition type. Example simple direct return types would be things such as i32, fp32. For compound types, a direct return type must be fully declared. Example fully defined types: VARCHAR(20), DECIMAL(25,5)
+
+### Return Type Expressions
+
+Any function can declare a return type expression. A return type expression uses a simplified set of expressions to describe how the return type should be returned. For example, a return expression could be as simple as the return of parameter declared in the arguments. For example `f(List<K>) =>K ` or can be a simple mathematical or conditional expression such as `add(decimal(a,b), decimal(c,d)) => decimal(a+c, b+d)`. For the simple expression language, there is a very narrow set of types:
+
+* Integer: 64 bit signed integer (can be a literal or a parameter value)
+* Boolean: True and False
+* Type: A Substrait type (with possibly additional embedded expressions)
+
+These types are evaluated using a small set of operations to to support common scenarios. Since it is possible that output derivation ultimately determines that binding is not allowed, a `not_bindable` operation is provided that states that the particular binding is now sllowed. List of valid operations:
+
+```
+Math: +, -, *, /, min, max
+Boolean: &&, ||, !, <, >, ==
+Input Parameters: type parameter, literal parameter
+Literals: integer, type
+Special: !bindable
+```
+
+Fully defined with argument types:
+
+* `not(boolean x) => boolean` 
+* `and(boolean a, boolean b) => boolean` 
+* `or(boolean a, boolean b) => boolean` 
+* `multiply(integer a, integer b) => integer`
+* `divide(integer a, integer b) => integer`
+* `add(integer a, integer b) => integer`
+* `substract(integer a, integer b) => integer`
+* `min(integer a, integer b) => integer`
+* `max(integer a, integer b) => integer`
+* `equal(integer a, integer b) => boolean`
+* `greater_than(integer a, integer b) => boolean`
+* `less_than(integer a, integer b) => boolean`
+* `equal(Type a, Type b) => boolean`
+* `if(boolean a) then (integer) else (integer)`
+* `if(boolean a) then (type) else (type)`
+* `not_bindable()`
+
+#### Example Type Expressions Uses
+
+For reference, here are are some common output type derivations and how they can be expressed with a return type expression:
+
+| Operation                                     | Definition                                                   |
+| --------------------------------------------- | ------------------------------------------------------------ |
+| Add item to list                              | `add(<List<T>, T>) => List<T>)`                              |
+| Decimal Division                              | `divide(Decimal(P1,S1), Decimal(P2,S2)) => Decimal(P1 -S1 + S2 + MAX(6, S1 + P2 + 1), MAX(6, S1 + P2 + 1))` |
+| Do regex on only string maps to return values | `extract_values(Map<K,V>) => if(K==STRING) THEN LIST<V> ELSE !bindable` |
```

Comment:
Why wouldn't this just be written as `extract_values(Map<String, V>)`?

Reply:
Fair point, my original plan was to do multiple if statements. E.g. k==string or k==varchar. I just got lazy.

Reply:
Updated

Reply:
Maybe I missed it but is the plan to keep this as simple as possible? So, for example, if someone wanted to write a derivation function that corresponded to "any non-compound type", they're in for a giant `or` expression.

Reply:
My goal is that the language should be as simple as reasonable. I'm open to adding a in-list type idea but I'm not entirely convinced it is necessary. All non-compound types doesn't actually seem like a real thing. Numeric types probably seems more real.

Reply:
> All non-compound types doesn't actually seem like a real thing.

Well, I did base that example on extracting values from a map, and I think you'd probably at least want any non-compound type to be able to be part of a a map. :shrug: I don't feel too strongly about it, keeping it simple to start seems like a good idea.

Reply:
Hmm... not super important at this moment but I wonder if we're using compound types to mean different things. I meant compound types as done in substrait where it is any type that is parameterized (including things like decimal, varchar, etc). Using that definition, I think most people would probably want to treat string and varchar much the the same and could imagine both being used as a key to map.

I better understand your comment about what maps would be restricted to if you limit "compound" to be types that contain other types such as struct, map and list. I've generally been referring to that set as "complex" as opposed to "compound". (Happy to revise terminology if something else is better for people.)

Reply:
Ah yes! Sorry. I was definitely referring to what you've called _complex_ types, not _compound_ types.

Does that change your response at all? Fine if not, just wondering if you think "all non-complex types" is a predicate that will be commonly desired and used.

Reply:
My general inclination would be to expose a generic OR_EQUALITY_LIST(value, type, type,...,type) operation as opposed to declare specific classes. Specific classes have the problem that if you then declared u8, for example, it wouldn't be in the "numeric" group. I'm torn on whether OR is enough or OR_EQUALITY_LIST is necessary. Love to hear your gut and others.

Reply:
Before I answer that, are we expecting that every consumer/producer provides an output derivation where they are required, or will there be a default implementation? I could imagine not needing to worry about `OR` vs `IN` vs `OR_EQUALITY_LIST` (`OR` would probably suffice) if we have the right defaults to avoid having to repeat the same giant `OR` expression in every producer/consumer.

Reply:
The derivation definition would be in the function signature yaml (or whatever format/name we give it), not the actual plan. In a function signature yaml, I'd be happy for that format to support named type expressions to normalize that kind of pattern (so it doesn't have be declared repetitively).

Reply:
I see. In that case I suspect `OR` would be fine to start. Seems like it'd be very clear pretty quickly if an `IN` style operator/function was needed.